pkg/asset/installconfig/vsphere/permissions.go

```diff
@@ -0,0 +1,386 @@
+	}
+	if platform.DefaultDatastore == "" {
+		if len(datastores) == 1 {
+			return datastores[0], nil
+		}
+		return nil, errors.New("multiple datastores found but no defaultDatastore defined in platform spec")
+	}
+	for _, ds := range datastores {
+		if ds.Name() == platform.DefaultDatastore {
+			return ds, nil
+		}
+	}
+	return nil, errors.New("specified defaultDatastore not found")
+}
+
+func getDataCenterFromPlatform(ctx context.Context, platform *vspheretypes.Platform, finder Finder) (*object.Datacenter, error) {
+	dataCenters, err := getDataCenters(ctx, finder)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to list datacenters")
+	}
+
+	if platform.Datacenter == "" {
+		if len(dataCenters) == 1 {
+			return dataCenters[0], nil
+		}
+		return nil, errors.New("multiple datacenters found but no datacenter defined")
+	}
+	for _, dc := range dataCenters {
+		name := strings.TrimPrefix(dataCenters[0].InventoryPath, "/")
+		if name == platform.Datacenter {
+			return dc, nil
+		}
+	}
+	return nil, errors.New("specified datacenter not found")
+}
+
+func getNetworkFromPlatform(ctx context.Context, client *vim25.Client, platform *vspheretypes.Platform, finder Finder, datacenter string, cluster string) (*vim25types.ManagedObjectReference, error) {
+	networks, err := GetClusterNetworks(ctx, finder, datacenter, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to retrieve networks")
+	}
+	if platform.Network == "" {
+		if len(networks) == 1 {
+			networkRef := networks[0].Reference()
+			return &networkRef, nil
+		}
+		return nil, errors.New("multiple networks found but not network defined in platform spec")
+	}
+
+	validNetworkTypes := sets.NewString(
+		"DistributedVirtualPortgroup",
+		"Network",
+		"OpaqueNetwork",
+	)
+
+	for _, network := range networks {
+		if validNetworkTypes.Has(network.Reference().Type) {
+			name, err := GetNetworkName(ctx, client, network)
+			if err != nil {
+				return nil, errors.Wrap(err, "unable to get network name")
+			}
+			if name == platform.Network {
+				networkRef := network.Reference()
+				return &networkRef, nil
+			}
+		}
+	}
+	return nil, errors.New("specified network not found")
+}
+
+func getFolderFromConfig(ctx context.Context, platform *vspheretypes.Platform, finder Finder) (*vim25types.ManagedObjectReference, error) {
+	if platform.Folder == "" {
+		return nil, errors.New("user defined folder not defined")
+	}
+	return getFolderReference(ctx, platform.Folder, finder)
+}
+
+func getFolderReference(ctx context.Context, path string, finder Finder) (*vim25types.ManagedObjectReference, error) {
+	folderObj, err := finder.Folder(ctx, path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "specified folder not found")
+	}
+	ref := folderObj.Reference()
+	return &ref, nil
+}
+
+func getManagedObjectReferences(ctx context.Context, client *vim25.Client, platform *vspheretypes.Platform, finder Finder) (map[permissionGroup]vim25types.ManagedObjectReference, error) {
```

Comment:
Should a negative test be added that covers this?

Reply:
yep, just added one.